src/tasks.coffee

```diff
@@ -0,0 +1,314 @@
+        task.completedDate = new Date()
+        logger.info "Round completed for rerun task ##{task._id} - Task completed"
+      else
+        task.status = result.status
+        logger.info "Round completed for rerun task ##{task._id} - Task has been #{result.status}"
+
+    task.save (err) -> callback err
+
+# Process a task.
+#
+# Tasks are processed in rounds:
+# Each round consists of processing n transactions where n is between 1 and the task's batchSize,
+# depending on how many transactions are left to process.
+#
+# When a round completes, the task will be marked as 'Queued' if it still has transactions remaining.
+# The next available core instance will then pick up the task again for the next round.
+#
+# This model allows the instance the get updated information regarding the task in between rounds:
+# i.e. if the server has been stopped, if the task has been paused, etc.
+processNextTaskRound = (task, callback) ->
+  promises = []
+  nextI = task.transactions.length - task.remainingTransactions
+
+  for transaction in task.transactions[nextI ... nextI+task.batchSize]
+    do (transaction) ->
+      defer = Q.defer()
+
+      rerunTransaction transaction.tid, task._id, (err, response) ->
+        if err
+          transaction.tstatus = 'Failed'
+          transaction.error = err
+          logger.error err
+        else if response?.status is 'Failed'
+          transaction.tstatus = 'Failed'
+          transaction.error = response.message
+          logger.error response.message
+        else
+          transaction.tstatus = 'Completed'
+
+        task.remainingTransactions--
+        defer.resolve()
+
+      transaction.tstatus = 'Processing'
+
+      promises.push defer.promise
+
+  task.save ->
+
+  (Q.all promises).then -> finalizeTaskRound task, callback
+
+
+rerunTransaction = (transactionID, taskID, callback) ->
+  rerunGetTransaction transactionID, (err, transaction) ->
+    return callback err if err
+
+    # setup the option object for the HTTP Request
+    Channel.findById transaction.channelID, (err, channel) ->
+      return callback err if err
+
+      logger.info "Rerunning #{channel.type} transaction"
+
+      if channel.type is 'http' or channel.type is 'polling'
+        rerunSetHTTPRequestOptions transaction, taskID, (err, options) ->
+          return callback err if err
+
+          # Run the HTTP Request with details supplied in options object
+          rerunHttpRequestSend options, transaction, (err, HTTPResponse) ->
+            return callback err, HTTPResponse
+
+
+      if channel.type is 'tcp' or channel.type is 'tls'
+        rerunTcpRequestSend channel, transaction, (err, TCPResponse) ->
+          return callback err if err
+
+          # Update original
+          ctx =
+            parentID : transaction._id
+            transactionId : transactionID
+            transactionStatus: TCPResponse.status
+            taskID : taskID
+
+          rerunMiddleware.updateOriginalTransaction ctx, (err) ->
```

Comment:
I don't quite get why the middleware needs to be called here, surely its called when rerunning each transaction ie. https://github.com/jembi/openhim-core-js/blob/master/src/koaMiddleware.coffee#L99-L100

I understand this may be copied code, but still it seems redundant.